### ISSUE TITLE
Under dart, different results with same iteration

### ISSUE THREAD

**Issue by BruceChen2017** [microsoft/LightGBM#2741]
```
# coding: utf-8
import lightgbm as lgb
import pandas as pd
from sklearn.metrics import mean_squared_error
import numpy as np

print('Loading data...')
# load or create your dataset
df_train = pd.read_csv('../regression/regression.train', header=None, sep='\t')
df_test = pd.read_csv('../regression/regression.test', header=None, sep='\t')

y_train = df_train[0]
y_test = df_test[0]
X_train = df_train.drop(0, axis=1)
X_test = df_test.drop(0, axis=1)

# create dataset for lightgbm
lgb_train = lgb.Dataset(X_train, y_train)
lgb_eval = lgb.Dataset(X_test, y_test, reference=lgb_train)

# specify your configurations as a dict
params = {
    'boosting_type': 'dart',
    'objective': 'regression',
    'metric': {'l2',},
    'num_leaves': 31,
    'learning_rate': 0.05,
    'feature_fraction': 0.9,
    'bagging_fraction': 0.8,
    'bagging_freq': 5,
    'verbose': 0
}

print('Starting training...')
ret = {}
gbm = lgb.train(params,
                lgb_train,
                num_boost_round=100,
                valid_sets=[lgb_eval],
                evals_result=ret,
                verbose_eval=20,
                )
best_iter = np.argmin(ret["valid_0"]["l2"])+1
print(best_iter, min(ret["valid_0"]["l2"]))
print('Saving model...')
# save model to file
gbm.save_model('model.txt')

print('Starting predicting...')
# predict
for i in list(range(0,101,20))[1:]:
    y_pred = gbm.predict(X_test, num_iteration=i)
    # eval
    print('The rmse of ' + str(i) + 'iter prediction is:', mean_squared_error(y_test, y_pred,squared=True))
```     

```
Loading data...
Starting training...
[20]	valid_0's l2: 0.201026
[40]	valid_0's l2: 0.189632
[60]	valid_0's l2: 0.202754
[80]	valid_0's l2: 0.197465
[100]	valid_0's l2: 0.200294
45 0.18831254838622288
Saving model...
Starting predicting... 
The rmse of 20iter prediction is: 0.2838051029618767
The rmse of 40iter prediction is: 0.27281945464281654
The rmse of 60iter prediction is: 0.2523540094321268
The rmse of 80iter prediction is: 0.22095508073752387
The rmse of 100iter prediction is: 0.20029426104747516
```   
Why with the same iteration except max iteration 100, l2 loss is different under valid set evaluation and predicted result.Here set for validation and prediction is the same

**Comment by trivialfis** [comment]
Maybe related.  We recently fixed an bug that prediction used for evaluation is computed with tree drop out.

**Comment by guolinke** [comment]
The trees are changing during training of dart, so this is expected.
If you want the identical prediction of one specific iteration, you can save model of that iteration too, not the final model. 
There is a possible fix for this, but it will also break the model file format. We can save the shrinkage rates of previous tress at every iteration, and load the corresponding ones for specific iteration  in prediction. I am not sure wether the fix os worthy or not.

@trivialfis thanks! Could you provide more details about your fix?

**Comment by trivialfis** [comment]
We added a parameter `training` to XGBoost's predict function, to distinguish whether the prediction value is used for training or not.  This also fixed custom objective with dart.
